### PR TITLE
Surface <-> "tensor pointer"

### PR DESCRIPTION
**Pull request by lferraz**
# Description

Added classes:
- `PySurfaceFromPtr()` allows to copy the data inside a "python pointer" into a surface
- `PySurfaceToPtr()` allows to copy the data inside a surface into a "python pointer"

<del>Added method `Forward()` in classes `PyNvDecoder` and  `PyFFmpegDemuxer ` to move forward `n` frames without decoding them. </del>

Added functionality to method `Flush()` in class `PyNvEncoder` in order to allow to flush frames one by one.

# Status

**Ready for review**

### CODE REVIEW

File: PyNvCodec/src/PyNvCodec.cpp [hunk 1]
```diff
@@ -220,6 +221,65 @@ bool PySurfaceDownloader::DownloadSingleSurface(shared_ptr<Surface> surface,
...
+PySurfaceFromPtr::PySurfaceFromPtr(uint32_t width, uint32_t height,
+                                   Pixel_Format format, uint32_t gpuID)
+    : height(height), width(width), gpuID(gpuID), outputFormat(format) {
+    surface.reset(Surface::Make(format, width, height,
+                                CudaResMgr::Instance().GetCtx(gpuID)));
+}
+
+shared_ptr<Surface> PySurfaceFromPtr::Execute(CUdeviceptr ptr) {
+  if (!ptr) {
+    return shared_ptr<Surface>(Surface::Make(outputFormat));
```
Comment (lferraz) [inline PyNvCodec/src/PyNvCodec.cpp:264]: should we return an exception:
```
  if (!ptr) {
    std::stringstream ss;
    ss << __FUNCTION__;
    ss << ":Source video frame has void CUDA device ptr.";
    throw std::runtime_error(ss.str());
  }
```

Comment (rarzumanyan) [reply PyNvCodec/src/PyNvCodec.cpp:264]: It depends on what behavior you want to achieve. Basically there are 3 options:

- Return nullptr which will be mapped to `None` in Python.
- Throw exception.
- Return empty surface.

Last two are both okay-ish IMO. 
I generally prefer not to throw exceptions if it doesn't crash the program & return empty surface instead.

**Review by rarzumanyan — COMMENTED** [review]
Could you make it a separate function instead?
Same logic but name it something like `FlushSinglePacket()`?
Then the regular `Flush()` function will simply call `FlushSinglePacket()` in loop while it returns packet.

**Review by rarzumanyan — COMMENTED** [review]
This can be substituted with `Seek()` functionality.

**Review by rarzumanyan — COMMENTED** [review]
If it's designed to exchange the data between VPF Surfaces and PyTorch Tensors maybe it shall be put into PytorchNvCodec.cpp?

**Comment by lferraz** [comment]
> This can be substituted with `Seek()` functionality.

I need to test the `Seek()`, however my experience with `seek` in OpenCV is that it tends to be extremely slow. In contrast, the `Forward()` method is very very fast.

**Comment by lferraz** [comment]
> If it's designed to exchange the data between VPF Surfaces and PyTorch Tensors maybe it shall be put into PytorchNvCodec.cpp?

In fact it is not only restricted to PyTorch. Probably you can also map tensors from other frameworks. 

Moreover, one of the advantages of this approach is that you only need to compile one lib. E.g. I had problems to compile the `PytorchNvCodec` because I needed to compile with the same compiler as PyTorch was compiled (at least is what I understood).

The only diff between the 2 approaches is the place where you define the container of the data (e.g. in PyTorch the Tensor), with the proposed approach is on the user side while with the current approach (`PytorchNvCodec`) it is done internally.

**Comment by rarzumanyan** [comment]
> 
> 
> > This can be substituted with `Seek()` functionality.
> 
> I need to test the `Seek()`, however my experience with `seek` in OpenCV is that it tends to be extremely slow. In contrast, the `Forward()` method is very very fast.

Seek itself isn't complex, demuxer just goes through the Acess Units which can be easily find by start code without any decoding / reconstruction.

Although codecs use reference frames for intra prediction, so proper `Seek` implementation goes to closest I frame in past and decodes all frames between closest I frame and desired frame "under the hood" to return fully reconstructed desired frame without errors. 

Same thing applies to `Forward` - you can't just skip some frames before desired one, because desired frame may use some of skipped frames for prediction.

**Comment by lferraz** [comment]
> > > This can be substituted with `Seek()` functionality.
> > 
> > 
> > I need to test the `Seek()`, however my experience with `seek` in OpenCV is that it tends to be extremely slow. In contrast, the `Forward()` method is very very fast.
> 
> Seek itself isn't complex, demuxer just goes through the Acess Units which can be easily find by start code without any decoding / reconstruction.
> 
> Although codecs use reference frames for intra prediction, so proper `Seek` implementation goes to closest I frame in past and decodes all frames between closest I frame and desired frame "under the hood" to return fully reconstructed desired frame without errors.
> 
> Same thing applies to `Forward` - you can't just skip some frames before desired one, because desired frame may use some of skipped frames for prediction.

Tomorrow I will check the `Seek()`. Thank you.

**Comment by rarzumanyan** [comment]
@lferraz 

Thanks a lot for finalizing this!
As soon as I fix the `Seek()` functionality I'll merge this PR.
I guess the frame pts difference is caused by AVFormat Context being flushed after `Seek` which affects demuxer.

**Comment by lferraz** [comment]
> As soon as I fix the `Seek()` functionality I'll merge this PR.

wait, I am going to solve now the issues we discuss last week. I will tell you something asap.

**Comment by rarzumanyan** [comment]
>I will tell you something asap.

Ok, will continue the investigation with `Seek` meanwhile.

**Comment by rarzumanyan** [comment]
@lferraz 

Please check out latest `seek_support_new` branch.
After I flush decoder upon seek, on my local machine it returns exactly the requested frames:
```python
nvDec.DecodeSingleFrame(rawFrameNV12, nvc.SeekContext(17), packet_data)
print('frame pts: ', packet_data.pts)
nvDec.DecodeSingleFrame(rawFrameNV12, nvc.SeekContext(17), packet_data)
print('frame pts: ', packet_data.pts)
nvDec.DecodeSingleFrame(rawFrameNV12, nvc.SeekContext(16), packet_data)
print('frame pts: ', packet_data.pts)
nvDec.DecodeSingleFrame(rawFrameNV12, nvc.SeekContext(15), packet_data)
print('frame pts: ', packet_data.pts)
nvDec.DecodeSingleFrame(rawFrameNV12, nvc.SeekContext(18), packet_data)
print('frame pts: ', packet_data.pts)
```

returns
```
1700
1700
1600
1500
1800
```

**Comment by lferraz** [comment]
> @lferraz
> 
> Please check out latest `seek_support_new` branch.
> After I flush decoder upon seek, on my local machine it returns exactly the requested frames:
> 
> ```python
> nvDec.DecodeSingleFrame(rawFrameNV12, nvc.SeekContext(17), packet_data)
> print('frame pts: ', packet_data.pts)
> nvDec.DecodeSingleFrame(rawFrameNV12, nvc.SeekContext(17), packet_data)
> print('frame pts: ', packet_data.pts)
> nvDec.DecodeSingleFrame(rawFrameNV12, nvc.SeekContext(16), packet_data)
> print('frame pts: ', packet_data.pts)
> nvDec.DecodeSingleFrame(rawFrameNV12, nvc.SeekContext(15), packet_data)
> print('frame pts: ', packet_data.pts)
> nvDec.DecodeSingleFrame(rawFrameNV12, nvc.SeekContext(18), packet_data)
> print('frame pts: ', packet_data.pts)
> ```
> 
> returns
> 
> ```
> 1700
> 1700
> 1600
> 1500
> 1800
> ```

I've createde an issue for all the things related with the `seek`.

**Comment by lferraz** [comment]
@rarzumanyan finally I finish the code!!!! 🎉

**Comment by rarzumanyan** [comment]
Great news, thanks you so much for your contribution!
As soon as we're done with `Seek` feature I'll merge it.

**Comment by rarzumanyan** [comment]
Hi @lferraz 

As I've started to merge the PR I've found that `PySurfaceFromPtr` and `PySurfaceToPtr` classes break the compatibility with multi-planar `Surface` formats such as `YUV420` because mentioned classes accept just single `CUdeviceptr` which is not the case for `YUV420` and `YCbCr` as they have 3 separate CUDA memory allocations (and hence 3 different `CUdeviceptr`) under the hood.

So I've removed those very classes and added `SurfacePlane::Import` and `SurfacePlane::Export` methods which basically allow for the very same functionality although they comply with VPF architecture where every `Surface` consists of 1 or more `SurfacePlane`.

Please check out the `pr_171` branch where your PR is combined with my latest edits. I'm not sure I'm following the PyTorch tensor permutation logic you do in `SamplePyTorch.py` file so please test this sample - on my machine it either produces broken frames or returns `CUDA_INVALID_VALUE` error depending on the tensor import method. When `PYTORCHNVCODEC` conversion method is used, it dices up 1 video frame into 9 smaller frames which are kinda "tiled" which means proper bytes are in the memory, I'm just not feeding the `SurfacePlane::Import` method with correct video frame dimensions as I'm not too familiar with PyTorch API.

**Comment by rarzumanyan** [comment]
@lferraz 

I had some doubts about the lifetime of `surface_tensor` object which type is declared here:
https://github.com/lferraz/VideoProcessingFramework/blob/e3e7264dfb87c56bb0fe4704dea8b69dab50ea94/SamplePyTorch.py#L57

Although I don't have too much of Python knowledge and assume your code is correct.
I've pushed a changes to `pr_171` branch with simpler `SamplePyTorch.py` sample which only exports RGB `Surface` to Pytorch tensor, divide the tensor and imports it back to VPF. It works fine, please take a look at it.

**Comment by lferraz** [comment]
Thanks for all your comments and help. I will review everything this afternoon :)

**Comment by lferraz** [comment]
Finally I could check the `import` and `export` methods, they work fine. 🎉

I feel this approach is a little bit less flexible than the previous one but at the same time is more robust, so no discussion, congrats for the feature and thank you for your effort.

My only doubt is the fact that we are passing the pitch to the `import` and `export` methods. I can see in the `SurfacePlane` class that the pitch is already known ([`MemoryInterface.hpp` line 91](https://github.com/NVIDIA/VideoProcessingFramework/blob/086d7e4c5ae3fbd754ce7ad8979a4d2e13e4c5d8/PyNvCodec/TC/inc/MemoryInterfaces.hpp#L91)) so I do not feel it must be required as a parameter. 

I imagine that for the non-interleaved approach we will get 3 different pointers using the method `PlanePtr()` and the pitch in that case will be `w`. 

Finally, I will upload a very small change in `SamplePyTorch.py`. I prefer to use tensors with shape (h, w, 3) instead of (h, w * 3). From a memory point of view is exactly the same but semantically is different.

I've not checked the `seek` function, I think this can be done after merging this feature. :)
Moreover I want to start discussing the feature I added in PR #186

**Comment by lferraz** [comment]
I created PR #189 with the small change in `SamplePyTorch.py`.

**Comment by lferraz** [comment]
@rarzumanyan I'be seen that the PR is merged!!! Congrats.

> My only doubt is the fact that we are passing the pitch to the `import` and `export` methods. I can see in the `SurfacePlane` class that the pitch is already known ([`MemoryInterface.hpp` line 91](https://github.com/NVIDIA/VideoProcessingFramework/blob/086d7e4c5ae3fbd754ce7ad8979a4d2e13e4c5d8/PyNvCodec/TC/inc/MemoryInterfaces.hpp#L91)) so I do not feel it must be required as a parameter.
> 
> I imagine that for the non-interleaved approach we will get 3 different pointers using the method `PlanePtr()` and the pitch in that case will be `w`.

What do you thing about this comment?

**Comment by rarzumanyan** [comment]
Hi @lferraz 

>What do you thing about this comment?

The `Import()` and `Export()` methods allows to pass raw CUDA memory pointers without prior knowledge of the allocation nature. One may simply `Export()` one VPF `Surface` and `Import()` it into another and that would be perfectly fine & healthy API usage. 

So we shall take allocation pitch into consideration because this is the nature of CUDA API.

**Comment by lferraz** [comment]
> The `Import()` and `Export()` methods allows to pass raw CUDA memory pointers without prior knowledge of the allocation nature. One may simply `Export()` one VPF `SurfacePlane` and `Import()` it into another and that would be perfectly fine & healthy API usage.

I forgot that those methods are quite generic. Thank you so much for your help and explanations.

**Pull request merged**